### PR TITLE
removed amp_func argument to add_eig_src method for EigenModeSource

### PR DESCRIPTION
**Pull request by Arjun-Khurana**
Reference issue: https://github.com/NanoComp/meep/issues/2379#issuecomment-1401324620

### CODE REVIEW

**Comment by smartalecH** [comment]
cc @hammy4815

**Comment by hammy4815** [comment]
Would this not prevent `amp_func` from being used at all in the case where an `EigenModeSource` lacks a `DiffractedPlanewave`? I'm not too familiar with how `amp_func` gets used under the hood, but  in `add_source`, `mp.Source` adds a volume source using `amp_func`:

https://github.com/NanoComp/meep/blob/d370f1aa603c5f9f74eae2690952e7a3b3d3cb0d/python/source.py#L153-L154

However, the `EigenModeSource` overloads the `add_source` method. The only other place where I see it fed into the C++ backend is here:

https://github.com/NanoComp/meep/blob/d370f1aa603c5f9f74eae2690952e7a3b3d3cb0d/python/source.py#L671-L674

Which is getting removed in this PR. But if it does get used somewhere else somehow, then I think this is fine as far as the new framework for `add_source` is concerned.

**Comment by Arjun-Khurana** [comment]
Yeah this commit was just my quick and dirty solution for getting rid of the segfault in my example, something like this I think preserves the behavior in the case that an `amp_func` is supplied for an `EigenModeSource`:

```python
if isinstance(self.eig_band, mp.DiffractedPlanewave):
    add_eig_src(self.amp_func, diffractedplanewave) if self.amp_func else add_eig_src(diffractedplanewave)
else:
    add_eig_src(self.amp_func) if self.amp_func else add_eig_src()
```

However I'm not sure an `EigenModeSource` should really ever have a user-specified `amp_func`, since I thought the point of this type of source is that the spatial profile is deduced from the waveguide structure using MPB. If that's the case, I think we can remove the reference to `amp_func` entirely in the overload of `add_source`, if not I believe the above logic should work.

**Comment by hammy4815** [comment]
This will probably work, and maybe we pull it for now. However, I'm wondering if we should fix the bigger issue going on, i.e. why after the adjoint run is the eigenmode source set up incorrectly with regards to `amp_func`? If we can instead patch the memory leak, it might be more optimal of a solution.

**Comment by stevengj** [comment]
> However I'm not sure an EigenModeSource should really ever have a user-specified amp_func, since I thought the point of this type of source is that the spatial profile is deduced from the waveguide structure using MPB.

The user-specified `amp_func` is *multiplied* by the MPB amplitude, e.g. if you want to add some Gaussian envelope to get a localized window of an eigenmode: https://github.com/NanoComp/meep/blob/d370f1aa603c5f9f74eae2690952e7a3b3d3cb0d/src/mpb.cpp#L254

**Comment by smartalecH** [comment]
The issue isn't with the `amp_func`, it's the NULL pointer. So let's just use #2394.

File: python/source.py [hunk 1]
```diff
@@ -663,15 +663,13 @@ def add_source(self, sim):
             self.eig_resolution,
             self.eig_tolerance,
             self.amplitude,
+            self.amp_func
         ]
-        add_eig_src = functools.partial(
-            sim.fields.add_eigenmode_source, *add_eig_src_args
-        )
 
         if isinstance(self.eig_band, mp.DiffractedPlanewave):
```
Comment (smartalecH) [inline python/source.py:669]: We don't need to change this, so let's punt to the simpler PR.

**Comment by smartalecH** [comment]
(we can always re-open if I'm wrong)

**Pull request closed without merging**